pkg/errdefs/wrap.go

```diff
@@ -36,3 +36,35 @@ func WithStack(err error) error {
 func Cause(err error) error {
 	return errors.Cause(err)
 }
+
+// not call callers() on every wrap, see https://github.com/pkg/errors/issues/75#issuecomment-574580408
+func hasStack(err error) bool {
+	if err == nil {
+		return false
+	}
+
+	type causer interface {
+		Cause() error
+	}
+
+	type tracer interface {
+		StackTrace() errors.StackTrace
+	}
+
+	e := err
+	for {
+		if _, ok := e.(tracer); ok {
+			return true
+		}
+
+		cause, ok := e.(causer)
+		if !ok {
+			return false
+		}
+
+		e = cause.Cause()
+		if e == nil {
+			return false
+		}
+	}
+}
```

Comment:
Two questions:

1. Is it efficient to execute this loop every time when `WithStack` is been called?
2. Is it possible to fall in dead loop?

Reply:
dead loop -> infinite loop 😄

Reply:
@YikSanChan updated

Reply:
- Is it efficient to execute this loop every time when WithStack is been called?

It is not efficient, but it will only be enabled in DEBUG mode

- Is it possible to fall in dead loop?

It will return when it unravels down to the error that does not implement cause and the error at the beginning, so it will not infinite loop

Reply:
Is it possible to have a cycle in the cause chain?

Reply:
like this?

```go
package main

import "fmt"

func main() {

	err1 := fmt.Errorf("error1")

	e1 := fmt.Errorf("E1: %w", err1)
	e2 := fmt.Errorf("E2: %w", e1)
	e3 := fmt.Errorf("E3: %w $w", e1, e2)
}
```

Reply:
Or this:

```go
e1 := fmt.Errorf("error1")
e2 := fmt.Errorf("E2: %w", e1)
e1 := fmt.Errorf("E1: %w", e2)
```

Reply:
no cycle, see the playgroup: https://go.dev/play/p/GKqxuTmLuy_z

Reply:
But by practice, I need to replace `casuer` with `unwrap` :cold_face:

Reply:
Great! `Errorf` deep should deep copy the original error instead of linking it using pointer. `Cause` in this pr works in the same way with `%w` and `unwrap`, right?

Reply:
It is not clear whether deep copy or copy pointer, I found it seems difficult to construct the ring when constructing the example.

The error constructed with %w implements the Unwrap interface, not Cause, so I'll replace it.